### PR TITLE
Generate complete html files

### PR DESCRIPTION
**Pull request by twiss**
WIP. Works best in Chrome. @joshua-s how do you normally debug in Firefox? `console.log()`s don't show up for me because of something similar to https://bugzil.la/1047646.

This isn't the approach from https://github.com/codexa/firetext/issues/245#issuecomment-65721543, because I'd already started then, sorry about that.

Proper diff: https://github.com/codexa/firetext/pull/246/files?w=1

### CODE REVIEW

**Comment by ferndot** [comment]
`console.log()`s generally appear for me, although `console.error()`s can be hidden. I checked the code, and it looks like an error occurs before `load()` (https://github.com/twiss/firetext/blob/complete-html/modules/editor/scripts/docIO.js#L39) is called, preventing it from running. I do not know the specifics, but I could investigate if you would like.

**Comment by twiss** [comment]
@joshua-s thanks for looking into it. So, I only just realized that a simpler way of testing Firetext in Firefox is to just connect it to Dropbox. Seems obvious I know (humor me), but maybe we should mention that in the documentation in "Get Coding" and/or the no storage connected dialog (e.g. "go to the settings to connect ...") or just add buttons for that in the dialog.

**Comment by ferndot** [comment]
@twiss that is a fantastic idea. I filed an issue for your second proposition (#251). Do you think that we should update the documentation in addition to adding a revised message?

**Comment by twiss** [comment]
It wouldn't hurt, but it's probably not necessary.

**Comment by twiss** [comment]
Not done yet, but the idea should be clear.

Unfortunately, Firefox takes every opportunity to complain (sometimes silently) that this is a hack. >90% of the changes in this PR are for Firefox.

Importantly, I couldn't get MessageChannel to work. Probably not a security regression, but I didn't look into it much. Commit before I switched away from it at https://github.com/twiss/firetext/tree/complete-html-messagechannel.

**Comment by twiss** [comment]
Rebased. Though Internet Explorer nukes all variables on `document.open`, so if we ever want compatibility with IE it's best to ditch this.

File: modules/editor/editor.html [hunk 1]
```diff
@@ -79,6 +88,81 @@
     border-right-width: 0;
   }
   </style>
+  <script id="content_script" type="await/content">
```
Comment (ferndot) [inline modules/editor/editor.html]: Why is this moved into the HTML file?

Comment (twiss) [inline modules/editor/editor.html]: For some reason things like `document.addEventListener` don't work in Firefox from `editor.js`.

Comment (ferndot) [inline modules/editor/editor.html]: That is strange. Is there any way that we could only include the `document.addEventListener` code in the file? IMO we should try to avoid inline scripts. Also, they are a CSP violation: https://developer.mozilla.org/Apps/Build/installable_apps_for_Firefox_OS/CSP#Applicable_CSP_Restrictions

Comment (twiss) [inline modules/editor/editor.html]: I moved it to its own file. Not sure why this does work though.

**Comment by ferndot** [comment]
I think that this is ready for merge.

**Comment by twiss** [comment]
There's one more bug here: if you open and close a document and open a second one, you get a stream of errors (nothing user-visible, though). I tried very briefly to fix that a while ago, I'll try again.

**Comment by twiss** [comment]
Almost done now.

The original plan in #239 was to add a `max-width` to html documents, and not the editor, which is what this PR attempts. In the meantime, #241 was merged. Do you still want to do the original plan?

**Comment by ferndot** [comment]
Yeah, let's go with the original plan. It enables greater flexibility IMO.

**Comment by twiss** [comment]
There're still two more bugs that I'm aware of:
- The style for night mode (https://github.com/codexa/firetext/blob/52fd168/modules/editor/scripts/editor_night.js#L3-7) is added to the html. Maybe I should wait for #267 before fixing this?
- In Chrome, if you click in the padding-left the resulting cursor is wrong by padding-left px. (This is less critical IMO.)

**Comment by twiss** [comment]
> The style for night mode (https://github.com/codexa/firetext/blob/52fd168/modules/editor/scripts/editor_night.js#L3-7) is added to the html. Maybe I should wait for #267 before fixing this?

Fixed anyway. Although I'm now confused as to how this was ever supposed to work, since nowhere is the background of the editor set to dark? (Maybe we broke it recently, or maybe it does work on FFOS?) Anyway, if you want to merge #273 it doesn't matter.

**Comment by ferndot** [comment]
So, I guess it doesn't matter then ^_^ I will review this presently!

**Comment by ferndot** [comment]
Looks good! Sorry for the delay...

**Comment by twiss** [comment]
Rebased.

**Comment by ferndot** [comment]
Oh, sorry :fearful:. I think my commit messed you up. Are you ready for me to merge this?

**Comment by twiss** [comment]
No problem, rebased. Yep!

**Comment by twiss** [comment]
Thanks!
